src/vs/workbench/contrib/mcp/common/mcpService.ts

```diff
@@ -0,0 +1,156 @@
+		@IContextKeyService contextKeyService: IContextKeyService
+	) {
+		super();
+
+		this.userCache = instantiationService.createInstance(McpServerMetadataCache, StorageScope.PROFILE);
+		this.workspaceCache = instantiationService.createInstance(McpServerMetadataCache, StorageScope.WORKSPACE);
+
+		const definitionsObservable = derived(reader => {
+			const collections = this._mcpRegistry.collections.read(reader);
+			return collections.flatMap(collectionDefinition => collectionDefinition.serverDefinitions.read(reader).map(serverDefinition => ({
+				serverDefinition,
+				collectionDefinition,
+			})));
+		});
+
+		const updateThrottle = this._store.add(new RunOnceScheduler(() => {
+			const definitions = definitionsObservable.get();
+
+			const nextDefinitions = new Set(definitions);
+			const currentServers = this._servers.get();
+			const nextServers: IMcpServer[] = [];
+			for (const server of currentServers) {
+				const match = definitions.find(d => defsEqual(server, d));
+				if (match) {
+					nextDefinitions.delete(match);
+					nextServers.push(server);
+				} else {
+					server.dispose();
+				}
+			}
+			for (const def of nextDefinitions) {
+				nextServers.push(instantiationService.createInstance(McpServer, def.collectionDefinition, def.serverDefinition, def.collectionDefinition.scope === StorageScope.WORKSPACE ? this.workspaceCache : this.userCache));
+			}
+
+			this._servers.set(nextServers, undefined);
+		}, 500));
+
+		// Throttle changes so that if a collection is changed, or a server is
+		// unregistered/registered, we don't stop servers unnecessarily.
+		this._register(autorun(reader => {
+			definitionsObservable.read(reader);
+			updateThrottle.schedule(500);
+		}));
+
+
+		const tools = this._register(new MutableDisposable());
+
+		this._register(autorunWithStore((reader, store) => {
+
+			const servers = this._servers.read(reader);
+
+			// TODO@jrieken wasteful, needs some diff'ing/change-info
+			const newStore = new DisposableStore();
+
+			tools.clear();
+
+			for (const server of servers) {
+
+				for (const tool of server.tools.read(reader)) {
+
+					const ctxKey = new RawContextKey<boolean>(`mcp.tool.${tool.id}.enabled`, true);
+					const ctxInst = contextKeyService.createKey<boolean>(ctxKey.key, true);
+					store.add(toDisposable(() => ctxInst.reset()));
+					store.add(autorun(reader => {
+						ctxInst.set(tool.enabled.read(reader));
+					}));
+
+					newStore.add(toolsService.registerToolData({
+						id: tool.id,
+						displayName: tool.definition.name,
+						modelDescription: tool.definition.description ?? '',
+						inputSchema: tool.definition.inputSchema,
+						when: ctxKey.isEqualTo(true),
+						tags: ['mcp', 'vscode_editing']
+					}));
+					newStore.add(toolsService.registerToolImplementation(tool.id, {
+
+						async prepareToolInvocation(parameters, token) {
+							return {
+								confirmationMessages: {
+									title: localize('aaa', "Run tool from {0}", server.definition.label),
```

Comment:
We should figure out how to have a nicer looking confirmation message, and auto-run at some point of course. You might want to fill in the localization ids at least

Reply:
Any problem with 3P extensions being able to call these? I suppose not

Reply:
> You might want to fill in the localization ids at least

I think nls ids aren't surfaces anywhere, afaik it's all array indicies and not even the translators see them. We only use them as dedupe-key (which could be inferred from the value and extra-args). That we still write them is debt and we could do what we have done for the API - just strings w/o ids

Reply:
> We should figure out how to have a nicer looking confirmation message, and auto-run at some point of course.

Enlighten me. Is there more/better API for this? Ideally something that works for all tool calls and not special-cases for MCP-tools vs normal tools

Reply:
No, I'm just thinking more along the lines of something like a "display name" for these tools, but maybe that doesn't exist in MCP?

Probably this confirmation should include the arguments too?